### PR TITLE
Add SpeechCommands train/valid/test split

### PR DESCRIPTION
**Pull request by vincentqb**
Add SpeechCommands train/valid/test split. This follows the convention established in [gtzan](https://github.com/pytorch/audio/blob/master/torchaudio/datasets/gtzan.py#L1014).

### CODE REVIEW

**Comment by vincentqb** [comment]
Merging this before release would simplify the [tutorial]()https://colab.research.google.com/drive/1yIiDkm4ixjNr6dg6oowwfxeDPf1NlXHP?authuser=1#scrollTo=aGauAPh_l9A).

File: test/torchaudio_unittest/datasets/speechcommands_test.py [hunk 1]
```diff
@@ -53,43 +53,58 @@ class TestSpeechCommands(TempDirMixin, TorchaudioTestCase):
 
     root_dir = None
     samples = []
+    train_samples = []
+    valid_samples = []
+    test_samples = []
 
     @classmethod
-    def setUp(cls):
+    def setUpClass(cls):
```
Comment (vincentqb) [inline test/torchaudio_unittest/datasets/speechcommands_test.py:61]: setUp is ran at every method call, and so setUpClass should be used instead as is done for the other dataset.

**Review by cpuhrsch — APPROVED** [review]

File: torchaudio/datasets/speechcommands.py [hunk 2]
```diff
@@ -81,6 +92,26 @@ def __init__(self,
...
+        if subset == "validation":
+            walker = validation_list
+        elif subset == "testing":
+            walker = testing_list
+        elif subset == "training":
+            walker = filter(
+                lambda w: not (w in validation_list or w in testing_list), walker
+            )
+
         self._walker = list(walker)
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:115]: `list` is no longer necessary and should be removed.

File: torchaudio/datasets/speechcommands.py [hunk 3]
```diff
@@ -81,6 +92,26 @@ def __init__(self,
...
+
+        if subset in ["training", "validation"]:
+            filepath = os.path.join(self._path, "validation_list.txt")
+            with open(filepath) as f:
+                validation_list = [os.path.join(self._path, l.strip()) for l in f.readlines()]
+
+        if subset in ["training", "testing"]:
+            filepath = os.path.join(self._path, "testing_list.txt")
+            with open(filepath) as f:
+                testing_list = [os.path.join(self._path, l.strip()) for l in f.readlines()]
```
[the remark this answers is not in the document]
Comment (mthrok) [reply torchaudio/datasets/speechcommands.py:104]: Also note that the call to `readlines` here is redundant. You can simply to `[os.path.join(self._path, l.strip()) for l in f]`

File: torchaudio/datasets/speechcommands.py [hunk 4]
```diff
@@ -48,13 +48,24 @@ class SPEECHCOMMANDS(Dataset):
             The top-level directory of the dataset. (default: ``"SpeechCommands"``)
         download (bool, optional):
             Whether to download the dataset if it is not found at root path. (default: ``False``).
+        subset (Optional[str]):
+            Select a subset of the dataset [None, "training", "validation", "testing"].
+            (default: ``None``)
```
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py:55]: As you point out, `training_list.txt` doesn't exist, it's defined to be what's not in validation or testing. The other two are defined by the dataset. Is that what you meant?

File: torchaudio/datasets/speechcommands.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: sure, I'm ok with your suggestion

File: torchaudio/datasets/speechcommands.py [hunk 5]
```diff
@@ -81,6 +92,26 @@ def __init__(self,
 
         walker = walk_files(self._path, suffix=".wav", prefix=True)
```
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: Fair point

File: torchaudio/datasets/speechcommands.py [hunk 6]
```diff
@@ -81,6 +92,26 @@ def __init__(self,
...
         walker = filter(lambda w: HASH_DIVIDER in w and EXCEPT_FOLDER not in w, walker)
```
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: That's technically outside the scope of this PR, but I've updated it anyway since we are moving away from generators.

File: torchaudio/datasets/speechcommands.py (hunk 2 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: same

File: torchaudio/datasets/speechcommands.py [hunk 7]
```diff
@@ -79,9 +92,22 @@ def __init__(self,
...
+                return [os.path.join(self._path, l.strip()) for l in f]
+
+        if subset == "validation":
+            self._walker = load_list(VALIDATION_LIST)
+        elif subset == "testing":
+            self._walker = load_list(TESTING_LIST)
+        elif subset == "training":
+            excludes = load_list(VALIDATION_LIST) + load_list(TESTING_LIST)
+            walker = walk_files(self._path, suffix=".wav", prefix=True)
+            self._walker = [w for w in walker if HASH_DIVIDER in w and EXCEPT_FOLDER not in w and w not in excludes]
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:107]: Looking way better than the previous. One more push.

Rather than parsing the all the files present in the file system first then filter only the interested, it's better to only parse the needed one.

Also, repeating the combination of `walk_files` and filtering `HASH_DIVIDER` in two clauses looks overwhelming. You can combine them into a much shorter code like the bellow.

```python
elif subset == "testing":
    ...
else:
    paths = [str(p) for p in Path(self._path).glob('*/*_no_hash_*.wav')]
        # no need to check EXCEPT_FOLDER because
        # none of the wav file in the directory match with `_no_hash_`
    if subset == "training":
        excludes = set(_load_lists("validation_list.txt", "testing_list.txt"))
        paths = [p for p in paths if p not in excludes]
    self._walker = paths

File: torchaudio/datasets/speechcommands.py (hunk 2 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: With the change mentioned [above](https://github.com/pytorch/audio/pull/966/#discussion_r512198533), this is indeed no longer needed. Removed.

File: torchaudio/datasets/speechcommands.py [hunk 8]
```diff
@@ -79,9 +92,22 @@ def __init__(self,
                     download_url(url, root, hash_value=checksum, hash_type="md5")
                 extract_archive(archive, self._path)
 
-        walker = walk_files(self._path, suffix=".wav", prefix=True)
-        walker = filter(lambda w: HASH_DIVIDER in w and EXCEPT_FOLDER not in w, walker)
-        self._walker = list(walker)
+        def load_list(filename):
+            filepath = os.path.join(self._path, filename)
+            with open(filepath) as fileobj:
+                return [os.path.join(self._path, line.strip()) for line in fileobj]
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:98]: Please prefer a plain function defined on module level especially when there is no local variable needs to be captured.

Also, since you are defining a helper function for your own sake, you can adopt a signature like

```
def _load_lists(*filename)
```

and it will make `excludes = load_lists(VALIDATION_LIST) + load_list(TESTING_LIST)` into as simple as `excludes = _load_lists("validation_list.txt", "testing_list.txt")`. (continue to the next comment)

**Review by mthrok — COMMENTED** [review]
Definitely better than the previous code. Made some comments to push it to the next level.

File: test/torchaudio_unittest/datasets/speechcommands_test.py [hunk 9]
```diff
@@ -53,43 +53,59 @@ class TestSpeechCommands(TempDirMixin, TorchaudioTestCase):
...
+                        )
+                        cls.samples.append(sample)
+                        if j < 2:
+                            cls.train_samples.append(sample)
+                        elif j < 4:
+                            label_filename = os.path.join(label, filename)
+                            valid.write(f'{label_filename}\n')
+                            cls.valid_samples.append(sample)
+                        elif j < 6:
+                            label_filename = os.path.join(label, filename)
```
Comment (mthrok) [inline test/torchaudio_unittest/datasets/speechcommands_test.py:106]: The contents are of `"testing_list.txt"` and `"validation_list.txt"` are in Linux format. 

```
$ head speech_commands_v0.01/testing_list.txt
bed/0c40e715_nohash_0.wav
bed/0ea0e2f4_nohash_0.wav
bed/0ea0e2f4_nohash_1.wav
```

Here, using `os.path.join` make it to the valid path on Windows system (something like `bed\0c40e715_nohash_0.wav`), so this is not correct. Something like the following is the correct way to mimic the data.

```suggestion
                            label_filename = f'{label}/{filename}'
```

Same goes to the case where `j < 4`

Comment (vincentqb) [reply test/torchaudio_unittest/datasets/speechcommands_test.py:106]: good catch

Comment (mthrok) [reply test/torchaudio_unittest/datasets/speechcommands_test.py:106]: I think it's more straightforward to use `List[pathlib.Path]` objects for `self._walker`, rather than converting them to `List[str]`.

File: torchaudio/datasets/speechcommands.py [hunk 10]
```diff
@@ -79,9 +92,22 @@ def __init__(self,
                     download_url(url, root, hash_value=checksum, hash_type="md5")
                 extract_archive(archive, self._path)
 
-        walker = walk_files(self._path, suffix=".wav", prefix=True)
-        walker = filter(lambda w: HASH_DIVIDER in w and EXCEPT_FOLDER not in w, walker)
-        self._walker = list(walker)
+        def load_list(filename):
+            filepath = os.path.join(self._path, filename)
+            with open(filepath) as f:
+                return [os.path.join(self._path, l.strip()) for l in f]
```
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: > Please prefer a plain function defined on module level especially when there is no local variable needs to be captured.

`load_list` uses `self._path` in its current form :)

> Also, since you are defining a helper function for your own sake, you can adopt a signature like
> 
> ```python
> def load_lists(*filename)
> ```
> 
> and it will make `excludes = load_list(VALIDATION_LIST) + load_list(TESTING_LIST)` into as simple as `excludes = load_list("validation_list.txt", "testing_list.txt")`. (continue to the next comment)

Either way looks reasonable to me, but I've made the update since you have a preference :)

File: torchaudio/datasets/speechcommands.py (hunk 7 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py:108]: I agree that parsing and then filtering is not optimal. Changing `os_walk` is out of scope for this PR though, and could change the order of the files from #814. I'm ok with replacing `os_walk` with `glob` in a next PR due to #950.

File: torchaudio/datasets/speechcommands.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mthrok) [reply torchaudio/datasets/speechcommands.py]: > `load_list` uses `self._path` in its current form :)

Oops my bad.

File: torchaudio/datasets/speechcommands.py (hunk 7 above)
Comment (mthrok) [reply torchaudio/datasets/speechcommands.py:108]: I encourage you to take this opportunity ro get rid of `walk_files` because, if you try to stick with `walk_files`, I do not see any pro at a glance. You will have to more works to settle with sub-optimal solution, which is not as compact yet inefficient.

Also #950 is about the support of `Path` object, so I do not see the logical connection between it and `walk_files` (which returns list of string).

Regarding, the order it returns, we have a unit test which can verify the proper ordering of files returned, so you do not need to worry about aligning the order with `walk_files` function. Instead, you can just focus on writing the correct code. However, this is a valid point and probably you will need to insert `sort` function in the snippet above.

File: torchaudio/datasets/speechcommands.py [hunk 11]
```diff
@@ -79,9 +90,25 @@ def __init__(self,
...
 
-        walker = walk_files(self._path, suffix=".wav", prefix=True)
-        walker = filter(lambda w: HASH_DIVIDER in w and EXCEPT_FOLDER not in w, walker)
-        self._walker = list(walker)
+        def _load_list(*filenames):
+            output = []
+            for filename in filenames:
+                filepath = os.path.join(self._path, filename)
+                with open(filepath) as fileobj:
+                    output += [os.path.join(self._path, line.strip()) for line in fileobj]
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:98]: I'm surprised our Windows CI job does not raise FileNotFound error with this implementation. If my understanding is correct, as previously noted, `os.path.join(self._path, line.strip())` would produce strings that look like `speech_commands\speech_commands_v0.01/testing_list.txt`, which is an invalid path on Windows.

File: test/torchaudio_unittest/datasets/speechcommands_test.py [hunk 12]
```diff
@@ -107,3 +121,51 @@ def testSpeechCommands(self):
             num_samples += 1
 
         assert num_samples == len(self.samples)
+
+    def testSpeechCommandsSubsetTrain(self):
+        dataset = speechcommands.SPEECHCOMMANDS(self.root_dir, subset="training")
```
Comment (mthrok) [inline test/torchaudio_unittest/datasets/speechcommands_test.py:126]: Can you add a test case for `subset=None`?

File: torchaudio/datasets/speechcommands.py (hunk 11 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: Yeah, that's what I see too. It seems like `os.path.normpath` would be useful here.

File: test/torchaudio_unittest/datasets/speechcommands_test.py (hunk 9 above)
Comment (vincentqb) [reply test/torchaudio_unittest/datasets/speechcommands_test.py]: Yeah, that will be more convenient to do that in the future

File: torchaudio/datasets/speechcommands.py [hunk 13]
```diff
@@ -79,9 +90,30 @@ def __init__(self,
...
+            self._walker = _load_list("validation_list.txt")
+        elif subset == "testing":
+            self._walker = _load_list("testing_list.txt")
+        elif subset == "training":
+            excludes = set(_load_list("validation_list.txt", "testing_list.txt"))
+            walker = walk_files(self._path, suffix=".wav", prefix=True)
+            self._walker = [
+                w for w in walker
+                if HASH_DIVIDER in w
+                and EXCEPT_FOLDER not in w
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:111]: `if HASH_DIVIDER in w and EXCEPT_FOLDER not in w` is redundant because files in background noises do not have `_nohash_` in filenames.

File: torchaudio/datasets/speechcommands.py [hunk 14]
```diff
@@ -79,9 +90,30 @@ def __init__(self,
                     download_url(url, root, hash_value=checksum, hash_type="md5")
                 extract_archive(archive, self._path)
 
-        walker = walk_files(self._path, suffix=".wav", prefix=True)
-        walker = filter(lambda w: HASH_DIVIDER in w and EXCEPT_FOLDER not in w, walker)
-        self._walker = list(walker)
+        def _load_list(*filenames):
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:93]: Can you move this to outside for the purpose of readability?

something like `def _load_list(root_dir, *filenames)`

File: torchaudio/datasets/speechcommands.py [hunk 15]
```diff
@@ -79,9 +90,30 @@ def __init__(self,
...
+                filepath = os.path.join(self._path, filename)
+                with open(filepath) as fileobj:
+                    output += [os.path.normpath(os.path.join(self._path, line.strip())) for line in fileobj]
+            return output
+
+        if subset == "validation":
+            self._walker = _load_list("validation_list.txt")
+        elif subset == "testing":
+            self._walker = _load_list("testing_list.txt")
+        elif subset == "training":
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:105]: I see there is a behavior inconsistency between `"training"/None` and `"validation"/"testing"`.

If certain valid files are removed from the dataset, then this dataset implementation will keep working for `"training"/None` case with less files, but for `"validation"/"testing"` it will raise an exception.

What's your take on that?

File: torchaudio/datasets/speechcommands.py (hunk 13 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py:112]: I'll avoid changing the logic here to avoid breaking any codes, say if a user moved a file in background_noise folder/etc. I can look into it for a later PR.

File: torchaudio/datasets/speechcommands.py (hunk 15 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py:106]: "testing", "validation", "training" are defined by the dataset by the training/validation files. It is technically undefined by the dataset outside of that. Given how the three are defined by the dataset in those files, as a user, I'd expect changes to those file to propagate. I'd say it'd be fair to add a quick note in the docstring.

Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py:107]: Added.

Comment (mthrok) [reply torchaudio/datasets/speechcommands.py:107]: > It is technically undefined by the dataset outside of that.

I agree with that, however IIRC, one of the goal of torchaudio's dataset implementation is to make it easy to modify the dataset. Something along the line of point 3&4 of https://github.com/pytorch/audio/issues/852#issuecomment-670590773 . With this rule, we have to take the extra step to think through what kind of modification is valid/invalid and what is the expected behavior, and put it in the implementation.

I am trying to raise the awareness of it.

Comment (mthrok) [reply torchaudio/datasets/speechcommands.py:107]: And of course, if we are just getting rid of the easy modification of the dataset, we do no need think about the any modification to dataset.

File: torchaudio/datasets/speechcommands.py (hunk 14 above)
Comment (mthrok) [reply torchaudio/datasets/speechcommands.py]: thanks this looks much readable now

File: torchaudio/datasets/speechcommands.py [hunk 16]
```diff
@@ -22,6 +22,15 @@
 }
 
 
+def _load_list(root, *filenames):
+    output = []
+    for filename in filenames:
+        filepath = os.path.join(root, filename)
+        with open(filepath) as fileobj:
+            output += [os.path.normpath(os.path.join(root, line.strip())) for line in fileobj]
```
Comment (mthrok) [inline torchaudio/datasets/speechcommands.py:30]: Can you add a comment on why `normapth` is necessary here?
Even though [`normpath`'s official document](https://docs.python.org/3/library/os.path.html#os.path.normpath) says `On Windows, it converts forward slashes to backward slashes. To normalize case, use normcase().`. For many of us, who are linux users, It is not well known. (I learned the only today.)

File: torchaudio/datasets/speechcommands.py (hunk 15 above)
Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py:107]: Actually, SpeechCommands does explain how those two files are generated, and how to generalize their approach, see [README](https://github.com/sankalp2610/Speech_Command_Recognition).

**Comment by vincentqb** [comment]
Merging based on offline discussion

File: torchaudio/datasets/speechcommands.py (hunk 7 above)
Comment (mthrok) [reply torchaudio/datasets/speechcommands.py]: @vincentqb Now that #950 is closed, can you followup replacing `walk_files` with `glob` with something like the following?

> I agree that parsing and then filtering is not optimal. Changing `os_walk` is out of scope for this PR though, and could change the order of the files from #814. I'm ok with replacing `os_walk` with `glob` in a next PR due to #950.

> ```python
> elif subset == "testing":
>     ...
> else:
>     paths = [str(p) for p in Path(self._path).glob('*/*_no_hash_*.wav')]
>         # no need to check EXCEPT_FOLDER because
>         # none of the wav file in the directory match with `_no_hash_`
>     if subset == "training":
>         excludes = set(_load_lists("validation_list.txt", "testing_list.txt"))
>         paths = [p for p in paths if p not in excludes]
>     self._walker = paths
> ```

Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: (about #940: #950 discussed using the Path library which arrived in 3.4. now that we use it, i'm ok with using other functions from it.)

Comment (vincentqb) [reply torchaudio/datasets/speechcommands.py]: good idea: i've opened an issue to request help from the community to remove `walk_files`: #1051.
